xcdat/averager.py

```diff
@@ -0,0 +1,458 @@
+"""Averager module that contains functions related to geospatial averaging.
+"""
+
+import xarray as xr
+import numpy as np
+
+# %%
+# Outline:
+#     spatial_average(): main spatial averaging function
+#         _validate_region_lon_bounds(): check/update regional longitude bounds so they are consistent with the variable array
+#             _switch_lon_bounds(): sub-function to actually switch the lon_bounds to a given type (either -180 to 180 or 0 to 360)
+#         _get_axis_weights(): get the default axis weights
+#         _combine_axis_weights(): combine axis weights into one weighting matrix
+#         _averager(): perform weighted average (with appropriate masking)
+
+# %% Notes
+
+# Differences between CDAT and this algorithm are smaller than 10**-4
+# CDAT is faster for smaller spatial scales, but this algorithm is faster for larger spatial scales
+# Using xarrays data reduction functionality gives similar results, but is slower than the averager function defined here
+# Averaging axes are specified with a list (e.g., axis=['lat', 'lon']) rather than the CDAT convention (axis='xy')
+# It would be helpful to be able to get axis weights (get_axis_weights()) more generically (some information is hardcoded)
+    # If .bounds had a method to return a list of bounds entries, we could create weights for all dimensions
+# May want to keep the axis weights or full weight array (i.e., add this information into the variable dataarray properties)
+# Is xvariable a good name?
+# Need to write unit tests
+# Can remove logging around bounds found in dataset
+# Functions are too specific to "lat" and "lon" (may have issues if dims are latitude / longitude)
+# Should remove these notes and outline...
+
+
+def _validate_region_lon_bounds(grid_lon_bounds, region_lon_bounds):
```

Comment:
Here's a refactored version that uses xarray syntax (pretty much the same as numpy)

https://github.com/tomvothecoder/xcdat/blob/1471ac8b4881379de37f9ea964596df708a745b8/xcdat/average_tv.py#L79-L133